src/Implementation/InputDefinition.php

```diff
@@ -72,19 +89,52 @@ public function isBoolean()
     }
 
     /**
-     * apply defaults and constraints and return cleaned array
-     * @param array $input
+     * @param mixed $value
+     * @return mixed
+     */
+    abstract protected function processValue($value);
+
+    /**
+     * @param string $value
+     * @return boolean
+     * @throws InvalidArgumentException
+     */
+    protected function matchesRegex($value)
+    {
+        if (!$this->regex) {
+            return true;
+        }
+        if (preg_match('/^'.$this->regex.'$/', "$value")) {
+            return true;
+        }
+        throw new InvalidArgumentException("$this->name doesn't match expected format '$this->regex'.");
+    }
+
+    /**
+     * @param mixed $value
      * @return array
-     * @throws \InvalidArgumentException if it's required or missing or doesn't match the constraints
      */
-    public function process($input)
+    private function getReturn($value)
     {
+        if ($value === null) {
+            return array();
+        }
+        return array($this->name => $value);
+    }
 
-        if(empty($input) && $this->isRequired()){
-            throw new \InvalidArgumentException("$this->name is required");
+    /**
+     * @param mixed $input
+     * @return array
+     * @throws InvalidArgumentException
+     */
+    final public function process($input)
+    {
+        if ((empty($input) || !is_array($input) || !array_key_exists($this->name, $input))) {
```

Comment:
ìs it possible to remove 1 pair of ( and )?

Reply:
yes, thanks for spotting that.